Add introduction to eBPF (#76)
canonical/ubuntu-server-documentation: explanation/intro-to/ebpf.md

(introduction-to-ebpf)=
# Introduction to eBPF

> eBPF is a lightweight virtual machine within the kernel. This virtual machine

Author: The closest we have is https://ebpf.io/ which is maintained by the eBPF community. It is listed at https://github.com/torvalds/linux/blob/master/MAINTAINERS#L4177, so it should be OK to link it here (done)

is exposed to the user-space and is able to safely run sandboxed programs in

> the kernel-space.

Author: Thanks, Sally! fixed.

BPF was originally designed to perform network packet filtering. Thus its name:
_Berkley Packet Filter_. Over time, BPF was extended to use more registers,
support 64-bit registers, data stores (Maps), and more. Such enhancement
allowed this extended version of BPF (eBPF) to be decoupled from the kernel
networking subsystem - eBPF was transformed in a tool with a broader scope,
capable of enhancing not only the networking experience, but also tracing,
profiling, observability, security, etc. Given its broader range of
applications, calling eBPF a packet filter no longer describes its full
capabilities. eBPF is now considered to be a standalone term and no longer
refers to _extended Berkley Packet Filter_, reflecting its expanded role. Also,

> the terms eBPF and BPF are used interchangeably, to refer to eBPF.

Author: Thanks, Sally!

I believe we missed the following information:

eBPF is now considered to be a standalone term and no longer
refers to _extended Berkley Packet Filter_

> the terms eBPF and BPF are used interchangeably, to refer to eBPF.

Reviewer: Updated my original suggestion - is this better?

> the terms eBPF and BPF are used interchangeably, to refer to eBPF.

Author: I think we lost the emphasis on why it does not make sense to call it a packet filter anymore. I applied your suggestions and added another sentence to reflect that. Let me know how it looks! :)

> the terms eBPF and BPF are used interchangeably, to refer to eBPF.

Reviewer: LGTM :) thanks!

> 

Reviewer: I'd maybe put a subheading in here like "how it works" or something, just to break up all the text

Author: Done. I added more sections after it to reflect what topic is being discussed.

User-space applications can load eBPF programs into the kernel as eBPF

> bytecode. Although you could write eBPF programs in bytecode, there are several

Author: This is what the kernel expects to be loaded (I tried to say that in the previous sentence, but maybe I failed to do so haha) Any suggestions here?

> bytecode. Although you could write eBPF programs in bytecode, there are several

Reviewer: I meant more like, why would you want to run programs in the kernel? (I'm trying to figure out if this is just a "I don't know" question vs. "our readers might not know")

> bytecode. Although you could write eBPF programs in bytecode, there are several

Author: It avoids context switching between user-space and kernel-space, which can be expensive. Hence the performance gains (we do discuss it later in the text). Not sure if we should change sth here. Thoughts?

> bytecode. Although you could write eBPF programs in bytecode, there are several

Reviewer: You also discussed it in the first paragraph - now I understand, and I think it's clear enough. I don't think any other changes are needed on this point, so marking resolved.

tools which provide abstraction layers on top of eBPF so you do not need to
write bytecode manually. These tools will then generate the bytecode which will
in turn be loaded into the kernel.

Once the eBPF program is loaded, it is then verified by the kernel before it
can run. These checks include verifying if the eBPF program halts and will
never get stuck in a loop, verifying that the program will not crash by
checking the registers and stack state validity throughout the program, and
ensuring the process loading the eBPF program has all the capabilities required

> by the eBPF program to run.

Reviewer: I'd put these three points into a bullet list, again to break up the text a bit and add visual interest

Author: Done!

After the verification step, the bytecode is Just-In-Time (JIT) compiled into
machine-code to optimize the program's execution.

> eBPF is a powerful tool for server and system administrators. It provides the

Reviewer: This feels a bit out of place after the "how it works" section, it's like we've gone back to the "who/what it's for" motivation, which should be near the top - I'd move this up to near the start (about line 5-6 I think?)

Reviewer: (with a little rewriting to avoid duplicating words on the next paragraph)

Author: Here, I added a new "eBPF in Ubuntu Server" section (this was originally the beginning of a section). Does this help?

Reviewer: The new title is good and suits the rest of the content, but I'd still prefer this first paragraph to be near the top (in the intro). It basically provides all the justification (the "why should I care?") that the reader is looking for to know whether it's relevant or interesting for them. Imo, once a reader is halfway down the page, they've already decided they're interested (and the "why" is a waste of their time at that point). If they don't get halfway down because we didn't provide a strong hook then they'll never see that this is relevant for them. This paragraph does a great job of explaining the "why", so I'd rather not bury it, but rather put it right in front of the reader's eyes :)

ability to extract plenty of contextual information from systems or processes,
offering several capabilities for monitoring, debugging, and securing systems
with little overhead given it runs in kernel-space, i.e., there is no need for
context-switching, making it very fast compared to solutions implemented in
user-space.

> 

Reviewer: I'd add another subheading around here like "how it can be used"

Author: Done (different heading title though)

> Namely, one can leverage tools like BCC (BPF Compiler Collection) or bpftrace

Reviewer: ```suggestion
Namely, one can leverage tools like BPF Compiler Collection (BCC) or `bpftrace`
```

Author: Fixed!

to identify bottlenecks, investigate performance degradation, trace specific
function calls, or create custom monitoring tools to collect data on specific
kernel or user-space processes without disrupting their services.

Since Ubuntu 24.04, `bpftrace` and `bpfcc-tools` (BCC) are available in every

> Ubuntu Server installation by default as part of our efforts to [enhance
> Ubuntu's application developers
> experience](https://discourse.ubuntu.com/t/spec-include-performance-tooling-in-ubuntu/43134).

Reviewer: ```suggestion
Ubuntu Server installation by default as part of our efforts to [enhance
the application developer's
experience](https://discourse.ubuntu.com/t/spec-include-performance-tooling-in-ubuntu/43134) in Ubuntu.
```

Author: Fixed!

`bpftrace` and `bpfcc-tools` install a range of tools with a variety of
different functionalities. Apart from the `bpftrace` tool itself, you can fetch
a comprehensive list of these tools with the following command:

```
$ dpkg -L bpftrace bpfcc-tools | grep -E '/s?bin/.*$' | xargs -n1 basename
```

Most of the tools listed above are quite well documented. Their manpages

> usually include good examples which you can try right away. The ones ended in

Reviewer: ```suggestion
usually include good examples you can try immediately. The tools ending in
```

Author: Fixed!

> `.bt` are installed by bpftrace and refer to bpftrace scripts (they are text

Reviewer: ```suggestion
`.bt` are installed by `bpftrace` and refer to `bpftrace` scripts (they are text
```

Author: Fixed!

files, hence, you could read them to understand how they are achieving specific

> tasks). The ones ended in `-bpfcc` are BCC tools (from bpfcc) written in Python

Reviewer: ```suggestion
tasks). The ones ending in `-bpfcc` are BCC tools (from `bpfcc`) written in Python
```

Author: Fixed

(you can also inspect those as you would inspect the `.bt` files).

You will also find several examples for `bpftrace` in
`/usr/share/doc/bpftrace/examples/` and for `bpfcc-tools` in
`/usr/share/doc/bpfcc-tools/examples/`.

For instance,

```
# bashreadline.bt
```

will print bash commands for all running bash shells in your system. Note that

> you need to run these scripts as root. You will also notice that there is also

Reviewer: ```suggestion
you need to run these scripts as root. You may notice that there is also
```

Author: Fixed

a `bashreadline-bpfcc` tool available from `bpfcc-tools`. Both of them provide
similar features. The former is implemented in python with BCC while the latter

> is a bpftrace script, as described above. You will see that many of these tools

Reviewer: ```suggestion
is a `bpftrace` script, as described above. You will see that many of these tools
```

Author: Fixed

have both a `-bpfcc` and a `.bt` version. Do read their manpages (and perhaps
the scripts) to choose which suits you best.

> On a more practical application, let's say you want to verify which bios is

Reviewer: ```suggestion
On a more practical application, let's say you want to verify which BIOS is
```

Author: Fixed

> used with particular commandlines of qemu. You could definitely use `strace` to

Reviewer: ```suggestion
used with a particular command line of QEMU. You could definitely use `strace` to
```

Author: Fixed

> do so, but that would likely add some overhead to the investigation process.

Reviewer: Why does it add more overhead?

Author: I added a comment to explain this

> Instead, you can simply use `opensnoop.bt` to trace `open()` syscalls:

Reviewer: ```suggestion
Instead, you can use `opensnoop.bt` to trace `open()` syscalls:
```

Author: Fixed

Reviewer: I think this one might have been missed

file-qemu-qboot.txt:1041:1.728519000   3413787 qemu-system-x86    12   0 /usr/share/seabios/vgabios-stdvga.bin
file-qemu-qboot.txt:1042:1.728626000   3413787 qemu-system-x86    12   0 /usr/share/seabios/vgabios-stdvga.bin
file-qemu-qboot.txt:1046:1.764491000   3413787 qemu-system-x86    12   0 /usr/share/terminfo/x/xterm-256color
file-qemu-seabios.txt:888:1.617490000   3413913 qemu-system-x86    12   0 /usr/share/seabios/bios-256k.bin
file-qemu-seabios.txt:889:1.617706000   3413913 qemu-system-x86    12   0 /usr/share/seabios/bios-256k.bin
file-qemu-seabios.txt:904:1.662015000   3413913 qemu-system-x86    12   0 /usr/share/qemu/kvmvapic.bin
file-qemu-seabios.txt:905:1.662068000   3413913 qemu-system-x86    12   0 /usr/share/qemu/linuxboot_dma.bin
file-qemu-seabios.txt:911:1.681084000   3413913 qemu-system-x86    12   0 /usr/share/seabios/vgabios-stdvga.bin
file-qemu-seabios.txt:912:1.681191000   3413913 qemu-system-x86    12   0 /usr/share/seabios/vgabios-stdvga.bin
file-qemu-seabios.txt:928:1.712340000   3413913 qemu-system-x86    12   0 /usr/share/terminfo/x/xterm-256color
```

Which shows one uses `qboot.rom` and the other uses `bios-256k.bin`.

> 

Reviewer: I'd put a subheading called "conclusion" or something like that

Author: Done.

eBPF offers a vast array of options to monitor, debug, and secure your systems
directly in kernel-space (i.e., fast and omniscient), with no need to disrupt
running services. It is an invaluable tool for system administrators and

> engineers.

Reviewer: What type of engineers?

Author: Fixed (hopefully)

For a complete documentation on eBPF, its internals and interfaces, please

> check the upstream documentation at https://docs.kernel.org/bpf/

Reviewer: ```suggestion
check the [upstream documentation](https://docs.kernel.org/bpf/)
```

Author: Fixed

(introduction-to-ebpf)=
# Introduction to eBPF

[eBPF](https://ebpf.io/) is a powerful tool for server and system
administrators. It provides the ability to extract plenty of contextual
information from systems or processes, offering several capabilities for
monitoring, debugging, and securing systems with little overhead given it runs

> in kernel-space, i.e., there is no need for context-switching, making it very

Author: I applied this change.  I have only used "on the other hand" to express ideas conflicting with the first one. This is the first time I see it applied to express a complimentary idea :)

> in kernel-space, i.e., there is no need for context-switching, making it very

Reviewer 2: Well, let me check with @s-makin if that is OK to be done that way :-)

it is usually more "this option has benefit A, the other has benefit B"
But it is defined as:
"used when you are comparing two different facts or two opposite ways of thinking about a situation"

So to me (might be wrong) it is also ok to say.
"Because of fact A it has benefit A', and because of fact B it has benefit B'"
as in
"On one hand due to fact A it has benefit A', and on the other hand because of fact B it has benefit B'"

@s-makin if this is stupid or even enough to be misleading it is easy to fix without involving Athos, let me know.

> in kernel-space, i.e., there is no need for context-switching, making it very

Reviewer 2: Sorted out as a team \o/

_Berkeley Packet Filter_. Over time, BPF was extended to use more registers,
support 64-bit registers, data stores (Maps), and more. Such enhancement
allowed this extended version of BPF (eBPF) to be decoupled from the kernel
networking subsystem - eBPF was transformed in a tool with a broader scope,
capable of enhancing not only the networking experience, but also tracing,
profiling, observability, security, etc. Given its broader range of
applications, calling eBPF a packet filter no longer describes its full
capabilities. eBPF is now considered to be a standalone term and no longer
refers to _extended Berkeley Packet Filter_, reflecting its expanded role. Also,
the terms eBPF and BPF are used interchangeably, to refer to eBPF.

## How it works

User-space applications can load eBPF programs into the kernel as eBPF
bytecode. Although you could write eBPF programs in bytecode, there are several
tools which provide abstraction layers on top of eBPF so you do not need to
write bytecode manually. These tools will then generate the bytecode which will
in turn be loaded into the kernel.

Once the eBPF program is loaded, it is then verified by the kernel before it

> can run. These checks include 

Reviewer: ```suggestion
can run. These checks include:
```

bytecode. Although you could write eBPF programs in bytecode, there are several
tools which provide abstraction layers on top of eBPF so you do not need to
write bytecode manually. These tools will then generate the bytecode which will
in turn be loaded into the kernel.

Once the eBPF program is loaded, it is then verified by the kernel before it
can run. These checks include:

* verifying if the eBPF program halts and will never get stuck in a loop,
* verifying that the program will not crash by checking the registers and stack

>   state validity throughout the program, and

Author: Leaving this to sally then

>   state validity throughout the program, and

Reviewer: There's no specific guidance for/against in the style guide, so it's really a personal preference thing. Personally, I like it and use it in my own writing (hence my lack of complaining) but if someone leaves it out I don't put it in. It's part of the flavour of an individual's writing style, and I leave that alone where possible.

>   state validity throughout the program, and

Reviewer 2: Discussed, we will keep it

eBPF is a powerful tool for server and system administrators. It provides the
ability to extract plenty of contextual information from systems or processes,
offering several capabilities for monitoring, debugging, and securing systems
with little overhead given it runs in kernel-space, i.e., there is no need for
context-switching, making it very fast compared to solutions implemented in
user-space.

Namely, one can leverage tools like the BPF Compiler Collection (BCC) or

> bpftrace to identify bottlenecks, investigate performance degradation, trace

Reviewer: ```suggestion
`bpftrace` to identify bottlenecks, investigate performance degradation, trace
```

```

Most of the tools listed above are quite well documented. Their manpages
usually include good examples you can try immediately. The tools ending in
`.bt` are installed by `bpftrace` and refer to `bpftrace` scripts (they are
text files, hence, you could read them to understand how they are achieving
specific tasks). The ones ending in `-bpfcc` are BCC tools (from `bpfcc`)
written in Python (you can also inspect those as you would inspect the `.bt`
files).

You will also find several examples for `bpftrace` in
`/usr/share/doc/bpftrace/examples/` and for `bpfcc-tools` in
`/usr/share/doc/bpfcc-tools/examples/`.

For instance,

```
# bashreadline.bt
```

will print bash commands for all running bash shells in your system. Note that
you need to run these scripts as root. You may notice that there is also a
`bashreadline-bpfcc` tool available from `bpfcc-tools`. Both of them provide
similar features. The former is implemented in python with BCC while the latter
is a `bpftrace` script, as described above. You will see that many of these
tools have both a `-bpfcc` and a `.bt` version. Do read their manpages (and
perhaps the scripts) to choose which suits you best.

> On a more practical application, let's say you want to verify which BIOS is

Reviewer 2: Andreas suggested to also add execsnoop which helped many of us a few times already.
Updated version here: https://dpaste.com/BJQEAGPM6

used with a particular command line of QEMU. You could definitely use `strace`
to do so, but that would likely add some overhead to the investigation

> process(`strace` uses `ptrace` and context switching may be required).

Reviewer: ```suggestion
process (since `strace` uses `ptrace`, context switching may be required).
```

file-qemu-seabios.txt:888:1.617490000   3413913 qemu-system-x86    12   0 /usr/share/seabios/bios-256k.bin
file-qemu-seabios.txt:889:1.617706000   3413913 qemu-system-x86    12   0 /usr/share/seabios/bios-256k.bin
file-qemu-seabios.txt:904:1.662015000   3413913 qemu-system-x86    12   0 /usr/share/qemu/kvmvapic.bin
file-qemu-seabios.txt:905:1.662068000   3413913 qemu-system-x86    12   0 /usr/share/qemu/linuxboot_dma.bin
file-qemu-seabios.txt:911:1.681084000   3413913 qemu-system-x86    12   0 /usr/share/seabios/vgabios-stdvga.bin
file-qemu-seabios.txt:912:1.681191000   3413913 qemu-system-x86    12   0 /usr/share/seabios/vgabios-stdvga.bin
file-qemu-seabios.txt:928:1.712340000   3413913 qemu-system-x86    12   0 /usr/share/terminfo/x/xterm-256color
```

Which shows one uses `qboot.rom` and the other uses `bios-256k.bin`.

## Conclusion

eBPF offers a vast array of options to monitor, debug, and secure your systems
directly in kernel-space (i.e., fast and omniscient), with no need to disrupt
running services. It is an invaluable tool for system administrators and
software engineers.

For a complete documentation on eBPF, its internals and interfaces, please

> check the [upstream documentation](https://docs.kernel.org/bpf/)

Reviewer: ```suggestion
check the [upstream documentation](https://docs.kernel.org/bpf/).
```
